can't build project using lib

Hello,

I'm trying to build a project using your library, and it does not compile with a couple of issues on the services.

Attached below the errors I'm getting when trying to build.

thelvaen@penguin:~/lpd_hub$ cargo tauri dev
        Info Watching /home/thelvaen/lpd_hub/src-tauri for changes...
    Blocking waiting for file lock on build directory
   Compiling oscq_rs v0.0.2
error[E0053]: method `call` has an incompatible type for trait
  --> /home/thelvaen/.cargo/registry/src/index.crates.io-6f17d22bba15001f/oscq_rs-0.0.2/src/service.rs:42:13
   |
42 |     fn call(&mut self, req: Request<IncomingBody>) -> Self::Future {
   |             ^^^^^^^^^
   |             |
   |             types differ in mutability
   |             help: change the self-receiver type to match the trait: `self: &OscQueryStatic`
   |
   = note: expected signature `fn(&OscQueryStatic, Request<_>) -> Pin<_>`
              found signature `fn(&mut OscQueryStatic, Request<_>) -> Pin<_>`

error[E0277]: the trait bound `tokio::net::TcpStream: hyper::rt::Read` is not satisfied
   --> /home/thelvaen/.cargo/registry/src/index.crates.io-6f17d22bba15001f/oscq_rs-0.0.2/src/service.rs:141:39
    |
141 |                     .serve_connection(stream, service)
    |                      ---------------- ^^^^^^ the trait `hyper::rt::Read` is not implemented for `tokio::net::TcpStream`
    |                      |
    |                      required by a bound introduced by this call
    |
    = help: the following other types implement trait `hyper::rt::Read`:
              &mut T
              Box<T>
              Upgraded
note: required by a bound in `hyper::server::conn::http1::Builder::serve_connection`
   --> /home/thelvaen/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-1.0.0-rc.4/src/server/conn/http1.rs:359:12
    |
353 |     pub fn serve_connection<I, S>(&self, io: I, service: S) -> Connection<I, S>
    |            ---------------- required by a bound in this associated function
...
359 |         I: Read + Write + Unpin,
    |            ^^^^ required by this bound in `Builder::serve_connection`

error[E0277]: the trait bound `tokio::net::TcpStream: hyper::rt::Write` is not satisfied
   --> /home/thelvaen/.cargo/registry/src/index.crates.io-6f17d22bba15001f/oscq_rs-0.0.2/src/service.rs:141:39
    |
141 |                     .serve_connection(stream, service)
    |                      ---------------- ^^^^^^ the trait `hyper::rt::Write` is not implemented for `tokio::net::TcpStream`
    |                      |
    |                      required by a bound introduced by this call
    |
    = help: the following other types implement trait `hyper::rt::Write`:
              &mut T
              Box<T>
              Upgraded
note: required by a bound in `hyper::server::conn::http1::Builder::serve_connection`
   --> /home/thelvaen/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-1.0.0-rc.4/src/server/conn/http1.rs:359:19
    |
353 |     pub fn serve_connection<I, S>(&self, io: I, service: S) -> Connection<I, S>
    |            ---------------- required by a bound in this associated function
...
359 |         I: Read + Write + Unpin,
    |                   ^^^^^ required by this bound in `Builder::serve_connection`

error[E0277]: the trait bound `tokio::net::TcpStream: hyper::rt::Read` is not satisfied
   --> /home/thelvaen/.cargo/registry/src/index.crates.io-6f17d22bba15001f/oscq_rs-0.0.2/src/service.rs:142:22
    |
142 |                     .await
    |                     -^^^^^
    |                     ||
    |                     |the trait `hyper::rt::Read` is not implemented for `tokio::net::TcpStream`
    |                     help: remove the `.await`
    |
    = help: the following other types implement trait `hyper::rt::Read`:
              &mut T
              Box<T>
              Upgraded
    = note: required for `hyper::server::conn::http1::Connection<tokio::net::TcpStream, OscQueryStatic>` to implement `Future`
    = note: required for `hyper::server::conn::http1::Connection<tokio::net::TcpStream, OscQueryStatic>` to implement `IntoFuture`

error[E0277]: the trait bound `tokio::net::TcpStream: hyper::rt::Write` is not satisfied
   --> /home/thelvaen/.cargo/registry/src/index.crates.io-6f17d22bba15001f/oscq_rs-0.0.2/src/service.rs:142:22
    |
142 |                     .await
    |                     -^^^^^
    |                     ||
    |                     |the trait `hyper::rt::Write` is not implemented for `tokio::net::TcpStream`
    |                     help: remove the `.await`
    |
    = help: the following other types implement trait `hyper::rt::Write`:
              &mut T
              Box<T>
              Upgraded
    = note: required for `hyper::server::conn::http1::Connection<tokio::net::TcpStream, OscQueryStatic>` to implement `Future`
    = note: required for `hyper::server::conn::http1::Connection<tokio::net::TcpStream, OscQueryStatic>` to implement `IntoFuture`

Some errors have detailed explanations: E0053, E0277.
For more information about an error, try `rustc --explain E0053`.
error: could not compile `oscq_rs` (lib) due to 5 previous errors

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,10 +12,11 @@ keywords = ["osc", "oscquery", "creative-coding"]
 homepage = "https://github.com/revilo196/oscq_rs"
 
 [dependencies]
-hyper = { version = "1.0.0-rc.3", features = ["full"] }
+hyper = { version = "1.0.0-rc.4", features = ["full"] }
 tokio = { version = "1", features = ["full"] }
 rosc = "0.10.0"
 serde = { version = "1.0.159", features = ["derive"] }
 serde_json = "1.0.95"
 serde_repr = "0.1"
 zeroconf = "0.10.5"
+pin-project-lite = "0.2.12"

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1,6 +1,7 @@
 mod oscquery_types;
 mod oscunit;
 mod service;
+mod tokiort;
 
 pub use oscquery_types::*;
 pub use oscunit::*;

--- a/src/service.rs
+++ b/src/service.rs
@@ -1,3 +1,4 @@
+use crate::tokiort::TokioIo;
 use crate::OSCNode;
 
 use hyper::server::conn::http1;
@@ -39,7 +40,7 @@ impl Service<Request<IncomingBody>> for OscQueryStatic {
     /// Handle an incoming HTTP request and return a future representing the eventual response.
     /// If the requested resource is not found, a 404 response is returned. If a query string is present,
     /// the appropriate response is generated based on the query. Otherwise, the full OSC query data is returned.
-    fn call(&mut self, req: Request<IncomingBody>) -> Self::Future {
+    fn call(&self, req: Request<IncomingBody>) -> Self::Future {
         // Create a response with the given string, including the appropriate "Content-Type" header.
         fn mk_response(s: String) -> Result<Response<String>, hyper::Error> {
             println!("{}", s);
@@ -134,11 +135,12 @@ pub async fn run_oscquery_service(
             let service = OscQueryStatic {
                 root: arc_root.clone(),
             };
+            let io = TokioIo::new(stream);
             tokio::task::spawn(async move {
                 println!("oscq_rs serve connection async {:?}", con);
                 if let Err(err) = http1::Builder::new()
                     .keep_alive(true)
-                    .serve_connection(stream, service)
+                    .serve_connection(io, service)
                     .await
                 {
                     println!("Failed to serve connection: {:?}", err);

--- a/src/tokiort.rs
+++ b/src/tokiort.rs
@@ -1,0 +1,150 @@
+#![allow(dead_code)]
+use pin_project_lite::pin_project;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+pin_project! {
+    #[derive(Debug)]
+    pub struct TokioIo<T> {
+        #[pin]
+        inner: T,
+    }
+}
+
+impl<T> TokioIo<T> {
+    pub fn new(inner: T) -> Self {
+        Self { inner }
+    }
+
+    pub fn inner(self) -> T {
+        self.inner
+    }
+}
+
+impl<T> hyper::rt::Read for TokioIo<T>
+where
+    T: tokio::io::AsyncRead,
+{
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        mut buf: hyper::rt::ReadBufCursor<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        let n = unsafe {
+            let mut tbuf = tokio::io::ReadBuf::uninit(buf.as_mut());
+            match tokio::io::AsyncRead::poll_read(self.project().inner, cx, &mut tbuf) {
+                Poll::Ready(Ok(())) => tbuf.filled().len(),
+                other => return other,
+            }
+        };
+
+        unsafe {
+            buf.advance(n);
+        }
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl<T> hyper::rt::Write for TokioIo<T>
+where
+    T: tokio::io::AsyncWrite,
+{
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        tokio::io::AsyncWrite::poll_write(self.project().inner, cx, buf)
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), std::io::Error>> {
+        tokio::io::AsyncWrite::poll_flush(self.project().inner, cx)
+    }
+
+    fn poll_shutdown(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        tokio::io::AsyncWrite::poll_shutdown(self.project().inner, cx)
+    }
+
+    fn is_write_vectored(&self) -> bool {
+        tokio::io::AsyncWrite::is_write_vectored(&self.inner)
+    }
+
+    fn poll_write_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &[std::io::IoSlice<'_>],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        tokio::io::AsyncWrite::poll_write_vectored(self.project().inner, cx, bufs)
+    }
+}
+
+impl<T> tokio::io::AsyncRead for TokioIo<T>
+where
+    T: hyper::rt::Read,
+{
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        tbuf: &mut tokio::io::ReadBuf<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        let filled = tbuf.filled().len();
+        let sub_filled = unsafe {
+            let mut buf = hyper::rt::ReadBuf::uninit(tbuf.unfilled_mut());
+
+            match hyper::rt::Read::poll_read(self.project().inner, cx, buf.unfilled()) {
+                Poll::Ready(Ok(())) => buf.filled().len(),
+                other => return other,
+            }
+        };
+
+        let n_filled = filled + sub_filled;
+        let n_init = sub_filled;
+        unsafe {
+            tbuf.assume_init(n_init);
+            tbuf.set_filled(n_filled);
+        }
+
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl<T> tokio::io::AsyncWrite for TokioIo<T>
+where
+    T: hyper::rt::Write,
+{
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        hyper::rt::Write::poll_write(self.project().inner, cx, buf)
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), std::io::Error>> {
+        hyper::rt::Write::poll_flush(self.project().inner, cx)
+    }
+
+    fn poll_shutdown(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        hyper::rt::Write::poll_shutdown(self.project().inner, cx)
+    }
+
+    fn is_write_vectored(&self) -> bool {
+        hyper::rt::Write::is_write_vectored(&self.inner)
+    }
+
+    fn poll_write_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &[std::io::IoSlice<'_>],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        hyper::rt::Write::poll_write_vectored(self.project().inner, cx, bufs)
+    }
+}

